overlord/devicestate/devicemgr.go

```diff
@@ -0,0 +1,384 @@
+	st.Lock()
+	defer st.Unlock()
+	// XXX: where to store device private key ultimately? use asserts support?
+	var encoded string
+	err := st.Get("device-keypair", &encoded)
+	if err == nil {
+		// nothing to do
+		return nil
+	}
+	if err != state.ErrNoState {
+		return err
+	}
+	keyPair, err := rsa.GenerateKey(rand.Reader, keyLength)
+	if err != nil {
+		return fmt.Errorf("cannot generate device key pair: %v", err)
+	}
+
+	encoded = base64.StdEncoding.EncodeToString(x509.MarshalPKCS1PrivateKey(keyPair))
+	st.Set("device-keypair", encoded)
+	return nil
+}
+
+func keyPair(st *state.State) (*rsa.PrivateKey, error) {
+	var encoded string
+	err := st.Get("device-keypair", &encoded)
+	if err == state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot find device key pair")
+	}
+	if err != nil {
+		return nil, err
+	}
+	b, err := base64.StdEncoding.DecodeString(encoded)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: cannot decode device key pair: %v", err)
+	}
+	pk, err := x509.ParsePKCS1PrivateKey(b)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: cannot decode device key pair: %v", err)
+	}
+	return pk, nil
+}
+
+type serialSetup struct {
+	Serial        string `json:"serial"`
+	SerialRequest string `json:"serial-request"`
+}
+
+type nonceTicketResp struct {
+	NonceTicket string `json:"nonce-ticket"`
+}
+
+var errPoll = errors.New("serial-request accepted, poll later")
+
+func submitSerialRequest(cli *http.Client, encodedSerialRequest []byte) (*asserts.Serial, error) {
+	resp, err := cli.Post(serialRequestURL, asserts.MediaType, bytes.NewBuffer(encodedSerialRequest))
+	if err != nil {
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+	if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+		return nil, fmt.Errorf("unexpected status code %d trying to post serial request to %s)", resp.StatusCode, serialRequestURL)
+	}
+
+	switch resp.StatusCode {
+	case 200, 201:
+	case 202:
+		return nil, errPoll
+	default:
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	// decode body with serial assertion
+	dec := asserts.NewDecoder(resp.Body)
+	got, err := dec.Decode()
+	if err != nil { // assume broken i/o
+		return nil, &state.Retry{After: 60 * time.Second}
+	}
+
+	serial, ok := got.(*asserts.Serial)
+	if !ok {
+		return nil, fmt.Errorf("serial-request response did not return a serial assertion")
+	}
+
+	return serial, nil
+}
+
+func doDeliverSerialRequest(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	devIdentity, err := auth.Device(st)
+	if err != nil {
+		return err
+	}
+
+	var retrieveSerialID string
+	err = t.Get("serial-setup-task", &retrieveSerialID)
+	if err != nil {
+		return err
+	}
+	retrieveSerial := st.Task(retrieveSerialID)
+	if retrieveSerial == nil {
+		return fmt.Errorf("internal error: cannot find retrieve serial task")
+	}
+
+	pk, err := keyPair(st)
+	if err != nil {
+		return err
+	}
+
+	cli := &http.Client{Timeout: 30 * time.Second}
```

Comment:
Can we name this as client?  I have a hard time not reading cli as command line interface.